File: packages/b2c-tooling-sdk/src/clients/scapi-fallback-backend.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
/**
 * Generic fallback wrapper for SCAPI/OCAPI dual backends.
 *
 * Builds a Proxy that implements the same interface as the underlying
 * backends. Each method call routes through {@link withFallback}: try SCAPI
 * first; on a recognized fallback trigger (e.g. `invalid_scope` or a
 * SCAPI-side capability gap such as a downgraded scope tier), fall back to
 * OCAPI for that call and pin to OCAPI for the rest of the wrapper's life.
 * Note that a successful SCAPI call only pins *softly* — a later call that
 * trips a fallback trigger still routes to OCAPI and re-pins, so a flow
 * that reads under a read-only scope and then needs to write isn't stuck.
 * The `name` property reflects the currently-active backend ('scapi' before
 * the first call resolves, then whichever survived).
 *
 * @module clients/scapi-fallback-backend
 */
import {getLogger} from '../logging/logger.js';
import {isFallbackTrigger, type BackendBase} from './scapi-backend-utils.js';

/**
 * Internal state shared by all method invocations on a Proxy. Holds the
 * resolved backend so that once SCAPI succeeds (or we've fallen back to
 * OCAPI), subsequent calls skip the SCAPI attempt.
 */
interface FallbackState<T extends BackendBase> {
  scapi: T;
  ocapi: T;
  domainName: string;
  resolved?: T;
}

/**
 * Wraps a SCAPI call with automatic OCAPI fallback on a recognized
 * fallback trigger ({@link isFallbackTrigger}).
 *
 * Standalone helper so the Proxy traps and any future direct callers share
 * one definition.
 */
async function withFallback<T extends BackendBase, R>(
  state: FallbackState<T>,
  fn: (backend: T) => Promise<R>,
): Promise<R> {
  // Once we've fallen back to OCAPI, stay there — there's no SCAPI surface to
  // re-attempt. Errors propagate from OCAPI directly.
  if (state.resolved === state.ocapi) {
    return fn(state.ocapi);
  }

  // Either unresolved (first call), or already pinned to SCAPI by a prior
  // successful call. We still try SCAPI, but a fallback-trigger error here
  // routes this call through OCAPI and re-pins. This handles the case where
  // a read succeeds (pinning SCAPI) and then a write fails because the
  // ScopeTierManager has been downgraded to read-only — the write should
  // still satisfy through OCAPI rather than throwing.
  const target = state.resolved ?? state.scapi;
  try {
    const result = await fn(target);
    if (!state.resolved) state.resolved = state.scapi;
    return result;
  } catch (error) {
    if (isFallbackTrigger(error)) {
      getLogger().info(`SCAPI ${state.domainName} unavailable for this operation, falling back to OCAPI`);
      state.resolved = state.ocapi;
      return fn(state.ocapi);
    }
    throw error;
  }
}

/**
 * Creates a fallback wrapper over `scapi` and `ocapi` backends.
 *
 * The returned object presents the same interface as `T`. Method calls are
 * intercepted: the first call tries SCAPI; on `invalid_scope` it falls back
 * to OCAPI. The choice is cached for the wrapper's lifetime.
 *
 * **Contract:**
 * - Both `scapi` and `ocapi` must implement `T`. TypeScript enforces this
 *   at the call site since both are typed as `T`.
 * - Only methods of `T` are routed through the fallback logic. The `name`
 *   getter is special-cased to reflect the resolved backend.
 * - **Non-method properties** are returned from the SCAPI target only and
 *   are not switched on fallback. By convention, backends should be method
 *   bags — any state beyond `name` should be encapsulated, not exposed.
 * - **Concurrency:** if two calls race before resolution, both may attempt
 *   SCAPI. This is benign for read operations (idempotent retries) and
 *   acceptable for writes (both either succeed or fail with the same
 *   error). Each Proxy instance has its own state, so this concerns only
 *   shared use of a single wrapper.
 *
 * @param scapi - Primary (SCAPI) backend implementation
 * @param ocapi - Fallback (OCAPI) backend implementation
 * @param domainName - Used in fallback log messages, e.g. `'jobs'`
 * @returns A Proxy over `scapi` whose methods route through fallback logic
 *
 * @example
 * ```ts
 * const backend = createFallbackBackend<JobsBackend>(scapiJobs, ocapiJobs, 'jobs');
 * await backend.executeJob('my-job'); // tries SCAPI, may fall back to OCAPI
 * ```
 */
export function createFallbackBackend<T extends BackendBase>(scapi: T, ocapi: T, domainName: string): T {
  const state: FallbackState<T> = {scapi, ocapi, domainName};

  return new Proxy(scapi, {
    get(target, prop, receiver) {
      // Special property: `name` reflects whichever backend has handled requests so far.
      if (prop === 'name') {
        return (state.resolved ?? scapi).name;
      }

      const value = Reflect.get(target, prop, receiver);

      // Non-functions (constants, getters): return as-is from the SCAPI backend.
      // Wrappers don't currently expose any non-method state besides `name`,
      // but this keeps the Proxy transparent for property access.
      if (typeof value !== 'function') {
        return value;
      }

      // For each method, return a wrapper that routes the call through fallback.
      // We must look up the method by name on the resolved backend (not on the
      // SCAPI target we're proxying), since the OCAPI backend may have a
      // different implementation.
      //
      // If the method is missing from OCAPI (e.g., a SCAPI-only capability like
      // delete), don't attempt a fallback — let SCAPI handle it directly.
      // The caller should use the type-guard pattern (e.g. supportsDeleteJobExecution)
      // to detect this before calling.
      const ocapiHasMethod = typeof (ocapi as unknown as Record<string | symbol, unknown>)[prop] === 'function';
      if (!ocapiHasMethod) {
        return (...args: unknown[]) => {
          const fn = (scapi as unknown as Record<string | symbol, unknown>)[prop];
          return (fn as (...a: unknown[]) => Promise<unknown>).apply(scapi, args);
        };
      }

      return (...args: unknown[]) =>
        withFallback(state, (backend) => {
          const fn = (backend as unknown as Record<string | symbol, unknown>)[prop];
          if (typeof fn !== 'function') {
            throw new TypeError(`Method ${String(prop)} is not a function on ${backend.name} backend`);
          }
          return (fn as (...a: unknown[]) => Promise<unknown>).apply(backend, args);
        });
    },
  }) as T;
}

File: python/b2c-tooling-sdk/src/b2c_tooling_sdk/clients/scapi_fallback_backend.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""Generic fallback wrapper for SCAPI/OCAPI dual backends.

Mirrors ``src/clients/scapi-fallback-backend.ts``. The TypeScript version builds
a ``Proxy`` that implements the same interface as the underlying backends. Python
has no ``Proxy``, so :func:`create_fallback_backend` returns a wrapper object whose
``__getattr__`` intercepts attribute access: async methods route through
:func:`_with_fallback` (try SCAPI first; on a recognized safe fallback trigger,
fall back to OCAPI for that call and re-pin), while the ``name`` property reflects
the currently-active backend.

A successful SCAPI call pins *softly* — a later call that trips a fallback trigger
still routes to OCAPI and re-pins, so a flow that reads under a read-only scope and
then needs to write isn't stuck. Once pinned to OCAPI, calls stay there.
"""

from __future__ import annotations

from typing import Any, TypeVar, cast

from b2c_tooling_sdk.clients.scapi_backend_utils import BackendBase, is_fallback_trigger
from b2c_tooling_sdk.logging import get_logger

T = TypeVar("T", bound=BackendBase)


class _FallbackBackend:
    """Wraps ``scapi`` + ``ocapi`` backends, routing method calls through fallback.

    Presents the same interface as the wrapped backends via ``__getattr__``.
    Each instance holds its own resolution state.
    """

    def __init__(self, scapi: BackendBase, ocapi: BackendBase, domain_name: str) -> None:
        self._scapi = scapi
        self._ocapi = ocapi
        self._domain_name = domain_name
        self._resolved: BackendBase | None = None

    @property
    def name(self) -> str:
        """The currently-active backend's name ('scapi' until a call resolves)."""
        return (self._resolved or self._scapi).name

    async def _with_fallback(self, method: str, args: tuple[Any, ...], kwargs: dict[str, Any]) -> Any:
        # Once we've fallen back to OCAPI, stay there — there's no SCAPI surface
        # to re-attempt. Errors propagate from OCAPI directly.
        if self._resolved is self._ocapi:
            return await getattr(self._ocapi, method)(*args, **kwargs)

        # Either unresolved (first call) or already pinned to SCAPI by a prior
        # successful call. Try SCAPI; a fallback-trigger error routes this call
        # through OCAPI and re-pins (handles read-then-write after a read-only
        # scope downgrade).
        target = self._resolved or self._scapi
        try:
            result = await getattr(target, method)(*args, **kwargs)
            if self._resolved is None:
                self._resolved = self._scapi
            return result
        except Exception as error:  # noqa: BLE001 - inspected by is_fallback_trigger
            if is_fallback_trigger(error):
                get_logger().info(f"SCAPI {self._domain_name} unavailable for this operation, falling back to OCAPI")
                self._resolved = self._ocapi
                return await getattr(self._ocapi, method)(*args, **kwargs)
            raise

    def __getattr__(self, name: str) -> Any:
        # __getattr__ only fires for attributes not found normally (so `name`,
        # the private fields, and methods defined on this class are unaffected).
        # Guard private/dunder names to avoid recursion before __init__ finishes.
        if name.startswith("_"):
            raise AttributeError(name)

        value = getattr(self._scapi, name)

        # Non-callables (constants, cached getters): return as-is from SCAPI.
        if not callable(value):
            return value

        # If the method is missing from OCAPI (a SCAPI-only capability), don't
        # attempt a fallback — call SCAPI directly. Callers should use a
        # capability guard before invoking such a method.
        if not callable(getattr(self._ocapi, name, None)):
            return value

        async def wrapper(*args: Any, **kwargs: Any) -> Any:
            return await self._with_fallback(name, args, kwargs)

        return wrapper


def create_fallback_backend(scapi: T, ocapi: T, domain_name: str) -> T:
    """Create a fallback wrapper over ``scapi`` and ``ocapi`` backends.

    The returned object presents the same interface as ``T``. Method calls are
    intercepted: the first call tries SCAPI; on a safe fallback trigger it falls
    back to OCAPI and re-pins. The ``name`` property reflects the resolved backend.

    :param scapi: Primary (SCAPI) backend implementation.
    :param ocapi: Fallback (OCAPI) backend implementation.
    :param domain_name: Used in fallback log messages, e.g. ``"jobs"``.
    :returns: A wrapper over ``scapi`` whose methods route through fallback logic.
    """
    return cast("T", _FallbackBackend(scapi, ocapi, domain_name))


__all__ = ["create_fallback_backend"]
